### ISSUE TITLE
`B023` False positive

### ISSUE THREAD

**Issue by tylerlaprade** [PyCQA/flake8-bugbear#380]
```python
arr = ['a', 'b', 'c', 'd', 'e', 'f']
mydict = {} 
for i in range(0, len(arr)):
    def foo():
        mydict[arr[i]] = 0
    foo()
```
This example is obviously too simple to be realistic - our actual code has more logic `foo()` and passes parameters to it, but is equivalent in terms of everything the rule cares about. We get an error of `Function definition does not bind loop variable 'i'.Flake8(B023)`. This should not happen because `foo` is not a closure. It's never called outside the loop, and execution is never delayed. `i` should always have the correct value.

The reason we aren't immediately doing `mydict[arr[i]] = 0` is because we have other logic that determines the parameters passed to `foo()`.

**Comment by cooperlees** [comment]
Would accept a PR that somehow dives into the function scope to see i is use within the loop.

Just out of interest, does passing `i` as a parameter to foo cause the check to no longer flag this check?

**Comment by tylerlaprade** [comment]
@cooperlees It does not, this code doesn't get a warning:
```
arr = ['a', 'b', 'c', 'd', 'e', 'f']
mydict = {} 
for i in range(0, len(arr)):
    def foo(x):
        mydict[arr[x]] = 0
    foo(i)
```
However this approach isn't feasible for our actual usecase.

**Comment by r-downing** [comment]
The warning is still technically correct about the binding issue in the function here. It might be out of scope for the checker to keep monitoring and see where the function is used?

Here's another workaround - bind the variable a default value:
```python
arr = ['a', 'b', 'c', 'd', 'e', 'f']
mydict = {} 
for i in range(0, len(arr)):
    def foo(x=i):
        mydict[arr[x]] = 0
    foo()
```

**Comment by jakkdl** [comment]
> It's never called outside the loop, and execution is never delayed.

This is somewhat tricky to check for. I think the only reasonable way to go about this would be to save `foo` as a potential B023 upon encountering it, and then raise the error if
1. `foo` is referenced outside the loop afterwards, or
2. `foo` is referenced in any way other than as a direct call (e.g. `blah(foo)`)

until the end of the parent scope where we clear it as a potential B023.

Glancing at check_for_b023 it's already quite complex though
